### PR TITLE
feat: Implement state action - sendExchangeAck status completed

### LINKED ISSUE #293: Implement state action - sendExchangeAck statuscompleted

### PR DESCRIPTION
**Pull request by talwinder50**
closes #293

### CODE REVIEW

**Comment by talwinder50** [comment]
So only thing left in the codedev patch is json,marshal test and decode string test. I am unable to produce error for these two scenarios. If anyone has suggestion please share.

File: pkg/didcomm/protocol/didexchange/states.go [hunk 1]
```diff
@@ -241,11 +242,24 @@ func (s *completed) Execute(msg dispatcher.DIDCommMsg, ctx context) (state, erro
 		if msg.Outbound {
 			return nil, fmt.Errorf("outbound responses are not allowed for state %s", s.Name())
 		}
-		// send ACK
+		response := &Response{}
+		err := json.Unmarshal(msg.Payload, response)
+		if err != nil {
+			return nil, err
```
Comment (rolsonquadras) [inline pkg/didcomm/protocol/didexchange/states.go]: better to return an error with a message. same for other cases. [Ref](https://github.com/hyperledger/aries-framework-go/blob/master/pkg/framework/context/context.go#L36).

File: pkg/didcomm/protocol/didexchange/states.go [hunk 2]
```diff
@@ -241,11 +242,24 @@ func (s *completed) Execute(msg dispatcher.DIDCommMsg, ctx context) (state, erro
...
 	case ConnectionAck:
-		// if msg.Outbound send ACK
-		// otherwise save did-exchange connection
+		if msg.Outbound {
+			err := ctx.handleOutboundAck(msg)
+			if err != nil {
+				return nil, err
+			}
+		}
+		//TODO: otherwise save did-exchange connection
```
Comment (rolsonquadras) [inline pkg/didcomm/protocol/didexchange/states.go]: Add the github issue# here.

Comment (talwinder50) [reply pkg/didcomm/protocol/didexchange/states.go]: added

**Comment by talwinder50** [comment]
The patch coverage is reduced to 78 percent because i edited the err message of untouched code. which are  basically marshal and decode string.

File: pkg/didcomm/protocol/didexchange/service_test.go [hunk 3]
```diff
@@ -10,10 +10,12 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"github.com/hyperledger/aries-framework-go/pkg/doc/did"
```
Comment (troyronda) [inline pkg/didcomm/protocol/didexchange/service_test.go]: should be grouped with the other Aries imports.

File: pkg/didcomm/protocol/didexchange/service_test.go [hunk 4]
```diff
@@ -14,6 +14,9 @@ import (
 	"testing"
 	"time"
 
+	"github.com/hyperledger/aries-framework-go/pkg/doc/did"
```
Comment (fqutishat) [inline pkg/didcomm/protocol/didexchange/service_test.go]: wrong location

File: pkg/didcomm/protocol/didexchange/states_test.go [hunk 5]
```diff
@@ -8,11 +8,15 @@ package didexchange
 
 import (
 	"encoding/json"
+	"fmt"
 	"testing"
 
+	"github.com/hyperledger/aries-framework-go/pkg/didcomm/protocol/decorator"
```
Comment (fqutishat) [inline pkg/didcomm/protocol/didexchange/states_test.go]: wrong location

File: pkg/framework/context/context_test.go [hunk 6]
```diff
@@ -12,6 +12,8 @@ import (
 	"fmt"
 	"testing"
 
+	"github.com/hyperledger/aries-framework-go/pkg/doc/did"
```
Comment (fqutishat) [inline pkg/framework/context/context_test.go]: wrong location

File: pkg/didcomm/protocol/didexchange/service_test.go (hunk 4 above)
Comment (talwinder50) [reply pkg/didcomm/protocol/didexchange/service_test.go]: done

File: pkg/framework/context/context_test.go (hunk 6 above)
Comment (talwinder50) [reply pkg/framework/context/context_test.go]: done

File: pkg/didcomm/protocol/didexchange/service_test.go (hunk 3 above)
Comment (talwinder50) [reply pkg/didcomm/protocol/didexchange/service_test.go]: done

File: pkg/framework/context/context_test.go (hunk 6 above)
Comment (talwinder50) [reply pkg/framework/context/context_test.go]: done

File: pkg/didcomm/protocol/didexchange/service_test.go (hunk 4 above)
Comment (talwinder50) [reply pkg/didcomm/protocol/didexchange/service_test.go]: done

File: pkg/didcomm/protocol/didexchange/service_test.go (hunk 3 above)
Comment (talwinder50) [reply pkg/didcomm/protocol/didexchange/service_test.go]: done

File: pkg/framework/context/context_test.go (hunk 6 above)
Comment (talwinder50) [reply pkg/framework/context/context_test.go]: done

File: pkg/didcomm/protocol/didexchange/states_test.go (hunk 5 above)
Comment (talwinder50) [reply pkg/didcomm/protocol/didexchange/states_test.go]: done

File: pkg/didcomm/protocol/didexchange/states.go [hunk 7]
```diff
@@ -412,3 +427,47 @@ func prepareConnectionSignature(connection *Connection) (*ConnectionSignature, e
...
+func (ctx *context) handleInboundResponse(response *Response) error {
+	ack := &model.Ack{
+		Type:   ConnectionAck,
+		ID:     uuid.New().String(),
+		Status: ackStatusOK,
+		Thread: &decorator.Thread{
+			ID: response.ID,
+		},
+	}
+	connBytes, err := base64.StdEncoding.DecodeString(response.ConnectionSignature.SignedData)
```
Comment (llorllale) [inline pkg/didcomm/protocol/didexchange/states.go]: This should be base64URL encoded - see [examples](https://github.com/hyperledger/aries-rfcs/tree/master/features/0023-did-exchange#example-1)

File: pkg/didcomm/protocol/didexchange/states.go [hunk 8]
```diff
@@ -412,3 +427,47 @@ func prepareConnectionSignature(connection *Connection) (*ConnectionSignature, e
...
+	if err != nil {
+		return err
+	}
+	conn := &Connection{}
+	err = json.Unmarshal(connBytes, conn)
```
Comment (llorllale) [inline pkg/didcomm/protocol/didexchange/states.go]: The first few bytes of `connBytes` contains a timestamp - see the examples in the RFC

Comment (talwinder50) [reply pkg/didcomm/protocol/didexchange/states.go]: Will be addressed in [issue-345](https://github.com/hyperledger/aries-framework-go/issues/345)

File: pkg/didcomm/protocol/didexchange/states.go (hunk 7 above)
Comment (talwinder50) [reply pkg/didcomm/protocol/didexchange/states.go]: changed

**Pull request merged**